### PR TITLE
feat: Add examples for Nutanix provider

### PR DESCRIPTION
**Pull request by deepakm-ntnx**
added nutanix examples for clusterclass and cluster

### CODE REVIEW

File: hack/examples/sync.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env bash
+#!/usr/bin/env bash -x
```
Comment (faiq) [inline hack/examples/sync.sh:1]:
```suggestion
#!/usr/bin/env bash 
```

Comment (deepakm-ntnx) [reply hack/examples/sync.sh:1]: any specific reason this should not be enabled by default?

Comment (deepakm-ntnx) [reply hack/examples/sync.sh:1]: https://www.shellcheck.net/wiki/SC2096

File: devbox.lock [hunk 2]
```diff
@@ -8,43 +8,15 @@
       "version": "1.6.27",
       "systems": {
         "aarch64-darwin": {
-          "outputs": [
```
Comment (jimmidyson) [inline devbox.lock:11]: Please revert this file - unless you're making specific changes to it to e.g. add/update/remove tools this should not be part of this PR. I think this is changing here because you have a different version of devbox installed, but that shouldn't affect your usage of this file.

File: hack/examples/patches/nutanix/initialize-variables.yaml [hunk 3]
```diff
@@ -17,26 +17,26 @@
             strategy:
```
Comment (jimmidyson) [inline hack/examples/patches/nutanix/initialize-variables.yaml:17]: These are coming from the initialize variables patch elsewhere so please just update `"/spec/topology/variables0/value/nutanix"`.

File: devbox.lock (hunk 2 above)
Comment (deepakm-ntnx) [reply devbox.lock]: reverted

File: hack/examples/patches/nutanix/initialize-variables.yaml (hunk 3 above)
Comment (deepakm-ntnx) [reply hack/examples/patches/nutanix/initialize-variables.yaml]: updated. thanks

**Comment by deepakm-ntnx** [comment]
@jimmidyson @faiq @dkoshkin @thunderboltsid any more comments? can we merge this?

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+  name: nutanix-quick-start-kcfg-0
+spec:
+  template:
+    spec:
+      joinConfiguration:
+        nodeRegistration:
+          kubeletExtraArgs:
+            cloud-provider: external
+            eviction-hard: nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<15%,memory.available<100Mi,imagefs.inodesFree<10%
+            tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
```
Comment (jimmidyson) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:15]: Why do we need to specify these here just for Nutanix provider? Aren't the defaults good enough? If not, then should we adopt this as a common set of ciphers across providers?

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    cluster.x-k8s.io/provider: nutanix
+  name: nutanix-quick-start-kcfg-0
+spec:
+  template:
+    spec:
+      joinConfiguration:
+        nodeRegistration:
+          kubeletExtraArgs:
+            cloud-provider: external
+            eviction-hard: nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<15%,memory.available<100Mi,imagefs.inodesFree<10%
```
Comment (jimmidyson) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:14]: Is this a good default for all providers? Should we make this a consistent default? I don't see why this would be specific to Nutanix.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+              tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+          scheduler:
+            extraArgs:
+              tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+        files:
+        - content: |
+            apiVersion: v1
+            kind: Pod
+            metadata:
+              name: kube-vip
```
Comment (jimmidyson) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:144]: kube-vip is going to be required across nutanix, vsphere, and pre-provisioned providers I think - should we make this a common patch to apply to our clusterclasses for the relevant providers?

File: examples/capi-quick-start/nutanix-cluster-calico-crs.yaml [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+  ca.crt: ${NUTANIX_ADDITIONAL_TRUST_BUNDLE=""}
+kind: ConfigMap
+metadata:
+  labels:
+    cluster.x-k8s.io/provider: nutanix
+  name: ${CLUSTER_NAME}-pc-trusted-ca-bundle
+---
+apiVersion: v1
+data:
+  nutanix-ccm.yaml: |
```
Comment (jimmidyson) [inline examples/capi-quick-start/nutanix-cluster-calico-crs.yaml:12]: CCM should be deployed as an addon with `AfterControlPlaneInitialized` hook - happy to do this as follow up PR though.

File: examples/capi-quick-start/nutanix-cluster-calico-crs.yaml [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+      name: nutanix-creds
+      namespace: kube-system
+    stringData:
+      credentials: |
+        [
+          {
+            "type": "basic_auth",
+            "data": {
+              "prismCentral":{
+                "username": "${NUTANIX_USER}",
```
Comment (jimmidyson) [inline examples/capi-quick-start/nutanix-cluster-calico-crs.yaml:270]: Same user for CCM as for cluster provisining used by CAPX - shouldn't these be separate and use less privileges (e.g. read-only) for CCM user?

**Review by jimmidyson — COMMENTED** [review]
Some comments but this is very close to being merged! Great work @deepakm-ntnx!

**Comment by faiq** [comment]
For now, I think it would be better to have non-working minimal examples here and later down the line adding the proper code in the form of GeneratePatch hooks and Lifecycle hooks to get started.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 4 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:15]: Tracking this separately to explore thru https://d2iq.atlassian.net/browse/D2IQ-100363

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 5 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:14]: Tracking this separately to explore thru https://d2iq.atlassian.net/browse/D2IQ-100363

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 6 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:144]: tracking thru https://d2iq.atlassian.net/browse/D2IQ-100364

File: examples/capi-quick-start/nutanix-cluster-calico-crs.yaml (hunk 7 above)
Comment (deepakm-ntnx) [reply examples/capi-quick-start/nutanix-cluster-calico-crs.yaml:12]: tracked thru https://d2iq.atlassian.net/browse/D2IQ-99841

File: examples/capi-quick-start/nutanix-cluster-calico-crs.yaml (hunk 8 above)
Comment (deepakm-ntnx) [reply examples/capi-quick-start/nutanix-cluster-calico-crs.yaml:270]: Good point. currently we have following user requirements
- CAPX: https://opendocs.nutanix.com/capx/v1.3.x/user_requirements/
- CCM: https://opendocs.nutanix.com/ccm/v0.3.x/requirements/
So technically we can allow having 2 different users configured for CCM and CAPX
will file a tracker to allow this on CAPX as well as CRE

**Comment by deepakm-ntnx** [comment]
> For now, I think it would be better to have non-working minimal examples here and later down the line adding the proper code in the form of GeneratePatch hooks and Lifecycle hooks to get started.

yes thats the plan if i understand your comment correctly. is it not the case in this PR?

**Review by dkoshkin — COMMENTED** [review]
Thanks for the tickets @deepakm-ntnx.
I'm good with merging this as is and working on those follow up tickets, but I will let @faiq and @jimmidyson do the actual approvals since they left the original comments.

**Comment by deepakm-ntnx** [comment]
> > For now, I think it would be better to have non-working minimal examples here and later down the line adding the proper code in the form of GeneratePatch hooks and Lifecycle hooks to get started.
> 
> yes thats the plan if i understand your comment correctly. is it not the case in this PR?

we will remove ccm related things from the cluster-topology example yaml once we move ccm installation as addon thru post cluster creation step in ccm https://d2iq.atlassian.net/browse/D2IQ-99841

**Review by faiq — DISMISSED** [review]
I strongly prefer to remove the CCM cluster resource sets from the examples. 

I can live with the rest of the stuff not being patches.

**Comment by deepakm-ntnx** [comment]
> I strongly prefer to remove the CCM cluster resource sets from the examples.
> 
> I can live with the rest of the stuff not being patches.

could you please add more details on why you feel so strongly? would it not be ok with you if we do it thru above mentioned tracker? i think this is really minor change and does not really add value. mainly https://d2iq.atlassian.net/browse/D2IQ-99841. given we are still in dev phase, we should run in iterations. cc @jimmidyson

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+                - name: kube-vip
+                  image: ghcr.io/kube-vip/kube-vip:v0.6.4
+                  imagePullPolicy: IfNotPresent
+                  args:
+                    - manager
+                  env:
+                    - name: vip_arp
+                      value: "true"
+                    - name: address
+                      value: "${CONTROL_PLANE_ENDPOINT_IP}"
```
Comment (jimmidyson) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:157]: This will not work as there is nothing templating this.

**Review by jimmidyson — DISMISSED** [review]

**Comment by faiq** [comment]
> While I would prefer CCM to be added as an addon at this point, to keep this moving and to make it possible to create a Nutanix cluster using CAREN hooks, I agree we should merge this once it's working and do the refactoring as a follow up. This will unblock other work, such as testing Nutanix CSI, etc, and allow adding e2e tests, etc.

We don't need CCM to be enabled to be able to test CSI.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 9 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:157]: its already being worked upon in https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/394/files#diff-3dda82929f9a5acd8a1ff63ccb56119ff0701f24ef2a8fcbb8fe0767f304d60aR138

Comment (jimmidyson) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:157]: But what's the point of merging this here then if it's not going to work?

Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:157]: 1. It will not break installation of helm chart and this cluster class does get created.
2. to merge this, are you suggesting we should make it work end to end? since that will require https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/6 to be merged along with all the capx PRs which is not really needed for these examples to be merged.
i am ok to wait till all these things are tested end to end and working. if thats the expectation.

**Comment by faiq** [comment]
> OK but CCM is required to bring a cluster fully up and make all nodes ready so it is still a blocker for e2e tests for Nutanix provider for example.

tests should fail until the code is complete.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 9 above)
Comment (jimmidyson) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:157]: OK maybe i'm being too strict... I would have liked a really basic CC to just work ootb and then add patches later to configure stuff, but I see your point so i'm happy to merge this as-is.

**Review by faiq — APPROVED** [review]

**Review by dkoshkin — APPROVED** [review]
